stm/src/types.rs

```diff
@@ -0,0 +1,224 @@
+        Self {
+            value: Some(MemoryController::new(var)),
+        }
+    }
+
+    /// Reads the value of the inner value without a transaction
+    /// FIXME: Do we really need this function,  or is this "only" required
+    /// for tests
+    pub fn read(&self) -> Result<Arc<T>, TransactionError> {
+        if let Some(ctrl) = &self.value {
+            match ctrl.value.read() {
+                Ok(lock) => return Ok(lock.clone()),
+                Err(error) => return Err(TransactionError::InconsistentState),
+            }
+        }
+
+        Err(TransactionError::InconsistentState)
+
+        // FIXME: this would require `BoxedMemory` to support safe cloning of memory
+        // Ok((*value).clone())
+    }
+
+    pub fn write(&self, value: T) -> Result<(), TransactionError> {
+        if let Some(ctrl) = &self.value {
+            match ctrl.value.write() {
+                Ok(mut lock) => {
+                    *lock = Arc::new(value);
+
+                    return Ok(());
+                }
+                Err(error) => return Err(TransactionError::InconsistentState),
+            }
+        }
+
+        Err(TransactionError::InconsistentState)
+    }
+}
+
+impl<T> Clone for TVar<T>
+where
+    T: Send + Sync + BoxedMemory,
+{
+    fn clone(&self) -> Self {
+        Self {
+            value: self.value.clone(),
+        }
+    }
+}
+
+impl<T> PartialEq for TVar<T>
+where
+    T: Send + Sync + BoxedMemory,
+{
+    fn eq(&self, other: &Self) -> bool {
+        let a = &self as *const _ as *const usize as usize;
+        let b = &other as *const _ as *const usize as usize;
+
+        // we compare only the pointers to store [`TVar`] inside a map
+        a == b
+    }
+}
+
+impl<T> PartialOrd for TVar<T>
+where
+    T: Send + Sync + BoxedMemory,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let a = &self as *const _ as *const usize as usize;
+        let b = &other as *const _ as *const usize as usize;
+
+        match a {
+            _ if a > b => Some(Ordering::Greater),
+            _ if a < b => Some(Ordering::Less),
+            _ => Some(Ordering::Equal),
+        }
+    }
+}
+
+impl<T> Hash for TVar<T>
+where
+    T: Send + Sync + BoxedMemory,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_usize((&self as *const _ as *const usize) as usize);
+        state.finish();
+    }
+}
+
+impl<T> Ord for TVar<T>
+where
+    T: Send + Sync + BoxedMemory,
+{
+    fn cmp(&self, other: &Self) -> Ordering {
+        let a = &self as *const _ as *const usize as usize;
+        let b = &other as *const _ as *const usize as usize;
+
+        match a {
+            _ if a > b => Ordering::Greater,
+            _ if a < b => Ordering::Less,
+            _ => Ordering::Equal,
+        }
+    }
+}
+
+impl<T> Eq for TVar<T> where T: Send + Sync + BoxedMemory {}
+
+/// Transactional Log type. The intend of this type
+/// is to track each operation on the target value
```

Comment:
this log would be encrypted right?  Don't remember what decision we came to on that one.

Reply:
yes. currently working on a solution how to do this transparently.